### PR TITLE
Lua module for Gossip protocol

### PR DESCRIPTION
**Pull request by alexandruantochi**
- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/*`.

\<Description of and rationale behind this PR\>

A module based on the Gossip protocol which allows the nodes to know the state of the network without relying on a single point of failure.

https://en.wikipedia.org/wiki/Gossip_protocol

### CODE REVIEW

**Comment by marcelstoer** [comment]
This fails the CI build. Do you have access to https://travis-ci.org/nodemcu/nodemcu-firmware/builds/639717733?

**Comment by alexandruantochi** [comment]
> This fails the CI build. Do you have access to https://travis-ci.org/nodemcu/nodemcu-firmware/builds/639717733?

Yep. I'll have a look to see what's wrong.

**Comment by alexandruantochi** [comment]
Fixed warnings from luacheck.

**Comment by marcelstoer** [comment]
> luacheck errors regarding unused variables

Yep, we have to thank @galjonsfigur for that 😃 👍

**Comment by alexandruantochi** [comment]
> > luacheck errors regarding unused variables
> 
> Yep, we have to thank @galjonsfigur for that 😃 👍

I had one unused variable because of an IF statement. 

I also had 2 that I was forced to use due to the [UDPsocket:on('receive', callback);](https://nodemcu.readthedocs.io/en/master/modules/net/#netudpsocketon) function. Maybe in the future we could have a way to ignore luacheck warnings for cases like this.

**Comment by HHHartmann** [comment]
I am not sure if I get the purpose correctly.
This is "only" to find other esps on the network? Would it also be possible to share some payload between the devices? Or am I missing something?

Would it make sense to identify the devices by their MAC address instead of the IP? So one would nit have to assign fixed IPs.

I like your coding style. Makes a nice separation of concerns.

Btw:
Luacheck allows for
```
-- luacheck: push no unused
local MCP23008_IODIR = 0x00
-- luacheck: pop
```

File: mkdocs.yml [hunk 1]
```diff
@@ -51,6 +51,7 @@ pages:
     - 'mcp23008': 'lua-modules/mcp23008.md'
     - 'redis': 'lua-modules/redis.md'
     - 'yeelink': 'lua-modules/yeelink.md'
+    - 'gossip': 'lua-modules/gossip.md'
```
Comment (HHHartmann) [inline mkdocs.yml]: Please sort alphabetically

File: lua_modules/gossip/gossip.lua [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+local gossip = {};
+local constants = {};
+local utils = {};
+local network = {};
+local state = {};
+
+-- Utils
+
+utils.debug = function(message)
+    if gossip.config.debug then
```
Comment (HHHartmann) [inline lua_modules/gossip/gossip.lua]: I think the coding style is 2 space indentation. Couldn't find a counter sample but I am not sure if it is documented.

**Comment by alexandruantochi** [comment]
@HHHartmann it's not for finding nodes on the network, it's for knowing their real time state (online or offline) and distributing this information between them.

A payload can be sent along with the mandatory `revision`, `heartbeat` and `state` values and this should be the user's own implementation.

e.g. sending this JSON works and the receiving node will have the extra param `payload`, which will be propagated to all the nodes in the network. Once the node sends another update with another payload, this information will be propagated as the heartbeat would be > 260

```JSON
{ "type" : "ACK", "192.168.0.158" : { "heartbeat" : 260, "revision" : 1, "state": 0, "payload" : "banana" } }
```

File: docs/lua-modules/gossip.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
+# ESPGossip
+
+| Since  | Origin / Contributor  | Maintainer  | Source  |
+| :----- | :-------------------- | :---------- | :------ |
+| 2020-01-20 | [alexandruantochi](https://github.com/alexandruantochi) | [alexandruantochi](https://github.com/alexandruantochi) | [gossip.lua](../../lua_modules/gossip/gossip.lua) |
+
+
+This module is based on the gossip protocol and usefull for use cases when multiple controllers are on the same network and they need to know about each other's state or when a single point of failure (such as an entity that queries each controller in particular) is to be avoided. It runs using UDP, thus providing faster communication and less network noise. 
+
```
Comment (HHHartmann) [inline docs/lua-modules/gossip.md]: Most documentation includes sections like 
## require
and
## release
followed by the functions.
There are however also other modules which don't follow this style.

**Comment by HHHartmann** [comment]
I alos wonder if it would make sense to allow for a callback on changed data which would be feed with `updatedNodes` from `updateNetworkState` or better with an array of these nodes (whole node or just the IP). Would also be helpful to have a sort of `push` when changing the `extra` information.

**Comment by alexandruantochi** [comment]
> I alos wonder if it would make sense to allow for a callback on changed data which would be feed with `updatedNodes` from `updateNetworkState` or better with an array of these nodes (whole node or just the IP). Would also be helpful to have a sort of `push` when changing the `extra` information.

New data is expected to come in continuously because each node will increment it's hearbeat before sending a `SYN` request. The idea is good, I could add a callback function that gets the `ip` and `nodeData` as parameters after the networkState has updated. I'll put it on my todo list.

**Comment by marcelstoer** [comment]
https://travis-ci.org/nodemcu/nodemcu-firmware/builds/645237431

**Comment by alexandruantochi** [comment]
> https://travis-ci.org/nodemcu/nodemcu-firmware/builds/645237431

I think everyone can see the link from the check, no need to bother adding it as a comment:
image

**Comment by marcelstoer** [comment]
I cannot know what you do and don't see. Previously it was suggested that non-committers wouldn't see the CI report link. Good to know this isn't the case (anymore).

**Comment by alexandruantochi** [comment]
Just trying to save you some trouble. Yeah, comment can be interpreted as me being irritated, sorry about that, it was not my intention  😆

**Comment by alexandruantochi** [comment]
Oh alright, I though your were adding it for me and wanted to save you some trouble.

File: lua_modules/gossip/gossip.lua [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+  setConfig = utils.setConfig,
+  start = state.start,
+  setRevFileValue = state.setRevFileValue,
+  networkState = {},
+  getNetworkState = utils.getNetworkState,
+  pushGossip = network.pushGossip
+};
+
+-- unit tests
+
```
Comment (HHHartmann) [inline lua_modules/gossip/gossip.lua]: Having to change the sources does not work well together with automated testing.
Using something like this together with a change in the test might help
``` Lua
-- load with loadfile("gossip.lua")("test") to run gossip_tests
if (... == "test") then
  gossip._test = {
   _gossip = gossip,
   _constants = constants,
   _utils = utils,
   _network = network,
   _state = state
  }
end
```

File: lua_modules/gossip/gossip_tests.lua [hunk 5]
```diff
@@ -0,0 +1,277 @@
+-- Gossip protocol implementation tests
+-- https://github.com/alexandruantochi/
+local gossipSubmodules = require('gossip');
+
```
Comment (HHHartmann) [inline lua_modules/gossip/gossip_tests.lua]: See above for the other part of the change.
``` Lua
local gossipSubmodules = loadfile("gossip.lua")("test");

local gossip = gossipSubmodules._test._gossip;
local constants = gossipSubmodules._test._constants;
local utils = gossipSubmodules._test._utils;
local network = gossipSubmodules._test._network;
local state = gossipSubmodules._test._state;
```

File: lua_modules/gossip/gossip_tests.lua [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+-- test constants and mocks
+
+local function dummy() return nil; end
+
+-- luacheck: push allow defined
+tmr = {};
+tmr.time = function() return 200; end
+sjson = {};
+sjson.decode = function(data) return data; end
+file = io;
```
Comment (HHHartmann) [inline lua_modules/gossip/gossip_tests.lua]: `file = {}` makes it run on device too

**Comment by HHHartmann** [comment]
Still found no time to test this but what I saw of the code looks good. So if you don't want to wait, this could go from my side.

**Comment by alexandruantochi** [comment]
Have a look first, I've learned from demos that the probability of failing is higher when somebody else is watching 😁

**Comment by alexandruantochi** [comment]
@marcelstoer what are the next steps I have to take to get this merged?

**Comment by marcelstoer** [comment]
Thanks for the nudge and thank you for your contribution. You don't need to do anything. Gregor already approved it a while ago. I guess he won't disapprove of me merging it then.

**Comment by HHHartmann** [comment]
Sorry for not reacting earlier. I was busy preparing vacation and planned to find some time on vacation but then went ill.
So now this is out to a bigger field experiment. 
But as i said before it all looks really good. Thanks for the middle.

**Comment by alexandruantochi** [comment]
> Sorry for not reacting earlier. I was busy preparing vacation and planned to find some time on vacation but then went ill.
> So now this is out to a bigger field experiment.
> But as i said before it all looks really good. Thanks for the module. .

I'm happy that you tested it and it performed well. I only had 4 modules available to run it on. Could you give me some details on your setup?

**Comment by HHHartmann** [comment]
Ah sorry. I didn't test it yet. It's still on my Todo list. I will also only test with 2 modules, so this probably will not add much.

**Pull request merged**